### PR TITLE
Add CompostItemEvent and EntityCompostItemEvent

### LINKED ISSUE #6349: CompostEvent
### Is your feature request related to a problem?

I wan't an event if a player compost something.

### Describe the solution you'd like.

An PlayerCompostEvent which is triggers if you rightclick the composter. A level method would be amazing.

### Describe alternatives you've considered.

PlayerInteractEvent should may work.

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### Other

_No response_

### PR DESCRIPTION
**Pull request by NoahvdAa**
Adds #6349 . The event exposes the player who composted, the composter block, the item stack used, whether or not the level will be raised (settable), the level the composter currently has and it is of course cancellable.

### CODE REVIEW

File: patches/server/0739-Add-PlayerCompostItemEvent.patch [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
++    // Paper start
++    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, boolean willRaiseLevel) {
++        // CraftBukkit start
++        return ComposterBlock.a(state, world, pos, item, willRaiseLevel);
++    }
++
+     static BlockState a(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand) {
+-        // CraftBukkit end
+         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
+         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
```
Comment (lynxplay) [inline patches/server/0739-Add-PlayerCompostItemEvent.patch]: Do you think there is any merit in moving the "will this itemstack in combination with a random value raise the compost level" logic into its own method to have a single place for the logic instead of repeating it ?

Comment (NoahvdAa) [reply patches/server/0739-Add-PlayerCompostItemEvent.patch]: I did it this way because they're only two lines, but an extra method would probably improve readability a bit.

Comment (lynxplay) [reply patches/server/0739-Add-PlayerCompostItemEvent.patch]: It might. Would also just ensure that the logic is updated (if it has to be) in all places that use it.
Obvious disadvantage is a bigger diff, so it really is something the opinion of others might be useful for.

File: patches/api/0327-Add-PlayerCompostItemEvent.patch [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
++     *
++     * @return the composter
++     */
++    @NotNull
++    public Block getComposter() {
++        return this.composter;
++    }
++
++    /**
++     * Gets the item that was used on  the composter.
```
Comment (aurorasmiles) [inline patches/api/0327-Add-PlayerCompostItemEvent.patch]:
```suggestion
+     * Gets the item that was used on the composter.
```

**Review by aurorasmiles — APPROVED** [review]
lgtm

**Comment by Machine-Maker** [comment]
This really seems like too much diff in the server patch for something so simple. Hold off on merging for now

**Comment by Machine-Maker** [comment]
Also, I feel like this event should not be player-specific, cause villagers also compost stuff, and hoppers can insert items as well.

File: patches/api/0327-Add-PlayerCompostItemEvent.patch (hunk 2 above)
Comment (NoahvdAa) [reply patches/api/0327-Add-PlayerCompostItemEvent.patch]: Whoops, that's my keyboard acting up. Should be fixed now.

File: patches/api/0327-Add-PlayerCompostItemEvent.patch [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when an item is about to be composted by a player.
++ */
++public class PlayerCompostItemEvent extends PlayerEvent implements Cancellable {
```
Comment (Machine-Maker) [inline patches/api/0327-Add-PlayerCompostItemEvent.patch]: Non-player's can also compost items, namely villagers and non-entities that insert items into the composter like hoppers.

File: patches/api/0327-Add-PlayerCompostItemEvent.patch [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
++    public void setWillRaiseLevel(boolean willRaiseLevel) {
++        this.willRaiseLevel = willRaiseLevel;
++    }
++
++    /**
++     * Gets the current composter level.
++     *
++     * @return the composter level
++     */
++    public int getLevel() {
```
Comment (Machine-Maker) [inline patches/api/0327-Add-PlayerCompostItemEvent.patch]: I feel like this is dangerous and could lead to de-sync (from the plugin author's perspective) between this value and the block's actual BlockData. What if an earlier event modifies the actual level of the composter?

File: patches/server/0739-Add-PlayerCompostItemEvent.patch [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+         return ComposterBlock.a(state, world, pos, item, world.getRandom().nextDouble());
+     }
+ 
++    // Paper start
++    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, boolean willRaiseLevel) {
++        // CraftBukkit start
++        return ComposterBlock.a(state, world, pos, item, willRaiseLevel);
++    }
++
+     static BlockState a(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand) {
```
Comment (Machine-Maker) [inline patches/server/0739-Add-PlayerCompostItemEvent.patch]: Rename this method and its usages with `build-data/mappings-patch.tiny`

File: patches/server/0739-Add-PlayerCompostItemEvent.patch [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+-        // CraftBukkit end
+         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
+         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
++        boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f); // Paper - Logic moved  up.
++        return ComposterBlock.a(iblockdata, generatoraccess, blockposition, itemstack, willRaiseLevel);
++    }
++
++    static BlockState a(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, boolean willRaiselevel) {
```
Comment (Machine-Maker) [inline patches/server/0739-Add-PlayerCompostItemEvent.patch]: `a` isn't an appropriate method name. No longer using obfuscated method names.

File: patches/server/0739-Add-PlayerCompostItemEvent.patch [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ 
+         if (i < 8 && ComposterBlock.COMPOSTABLES.containsKey(itemstack.getItem())) {
+             if (i < 7 && !world.isClientSide) {
+-                BlockState iblockdata1 = ComposterBlock.addItem(state, (LevelAccessor) world, pos, itemstack);
++                // Paper start
++                float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
++                boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && world.getRandom().nextDouble() >= (double) f); // Paper - Logic moved  up.
++
++                org.bukkit.Location composterLocation = net.minecraft.server.MCUtil.toLocation(world, pos);
++                com.destroystokyo.paper.event.player.PlayerCompostItemEvent event = new com.destroystokyo.paper.event.player.PlayerCompostItemEvent((org.bukkit.entity.Player) player.getBukkitEntity(), composterLocation.getBlock(), itemstack.getBukkitStack(), willRaiseLevel, i);
```
Comment (Machine-Maker) [inline patches/server/0739-Add-PlayerCompostItemEvent.patch]: Why is this here instead of in `addItem`? Moving it to `addItem` would negate the need to move the "random" logic up.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Any other methods in ComposterBlock that have obfuscated names should also be refactored with `build-data/mappings-patch.tiny`

File: patches/api/0327-Add-PlayerCompostItemEvent.patch (hunk 3 above)
Comment (NoahvdAa) [reply patches/api/0327-Add-PlayerCompostItemEvent.patch]: What would be the best way to do this? A CompostItemEvent as parent with `EntityCompostItemEvent` and `HopperCompostItemEvent` as children?

File: patches/api/0327-Add-PlayerCompostItemEvent.patch (hunk 4 above)
Comment (NoahvdAa) [reply patches/api/0327-Add-PlayerCompostItemEvent.patch]: Good catch, I'll remove this since the block can be casted to [Levelled](https://papermc.io/javadocs/paper/1.17/org/bukkit/block/data/Levelled.html) to get the level.

Comment (Machine-Maker) [reply patches/api/0327-Add-PlayerCompostItemEvent.patch]: Yeah, it’s hard to say what to do actually, because if you set the level that way, and don’t cancel the event, it will have practically no effect because it’s set after the event as well. Idk the best course of action here tbh.

File: patches/api/0327-Add-PlayerCompostItemEvent.patch (hunk 3 above)
Comment (Machine-Maker) [reply patches/api/0327-Add-PlayerCompostItemEvent.patch]: I don’t think you need 3 events, just at most 2 ItemCompostEvent, and EntityItemCompostEvent. You could probably just get away with one and have a nullable entity getter thats just null when it’s not coming from and entity’s action.

File: patches/server/0739-Add-PlayerCompostItemEvent.patch (hunk 7 above)
Comment (NoahvdAa) [reply patches/server/0739-Add-PlayerCompostItemEvent.patch]: This was moved up so the event could know whether or not the level would be raised. If I were to fire the event in `addItem`, I would have to move the code for decreasing the item, awarding a stat, etc. down aswell, this felt like a better choice.

Comment (Machine-Maker) [reply patches/server/0739-Add-PlayerCompostItemEvent.patch]: Take a look at [what I did here](https://gist.github.com/Machine-Maker/8053a26c555d8b300239dc0e6fc08bb5).

**Review by josephAttia — APPROVED** [review]

**Comment by NoahvdAa** [comment]
I've just pushed a commit that renames the event to `CompostItemEvent` (based on [MachineMaker's implementation](https://gist.github.com/Machine-Maker/8053a26c555d8b300239dc0e6fc08bb5)), which can now be triggered by entities (villagers/players) and hoppers.

File: patches/api/0327-Add-PlayerCompostItemEvent.patch (hunk 4 above)
Comment (NoahvdAa) [reply patches/api/0327-Add-PlayerCompostItemEvent.patch]: I've removed `getLevel()` for now, not sure if it should be added back.

File: patches/api/0327-Add-PlayerCompostItemEvent.patch (hunk 3 above)
Comment (NoahvdAa) [reply patches/api/0327-Add-PlayerCompostItemEvent.patch]: > You could probably just get away with one and have a nullable entity getter thats just null when it’s not coming from and entity’s action.

I've implemented EntityEvent and overriden `getEntity` to specify that it may be null in the javadocs, is this the intended way of adding extra javadoc?

File: patches/api/0327-Add-PlayerCompostItemEvent.patch [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.entity.EntityEvent;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * Called when an item is about to be composted by an entity or a hopper.
++ */
++public class CompostItemEvent extends EntityEvent implements Cancellable {
```
Comment (Machine-Maker) [inline patches/api/0327-Add-PlayerCompostItemEvent.patch]: This can't extend `EntityEvent` because the entity is nullable.
Look at the super constructor you are calling, it's entity param is marked as not null.

File: patches/server/0739-Add-PlayerCompostItemEvent.patch [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                 this.changed = true;
+-                BlockState iblockdata = ComposterBlock.addItem(this.state, this.level, this.pos, itemstack);
++                // Paper start
++                if (this.composterState == null) {
++                    return;
++                }
+ 
+-                this.level.levelEvent(1500, this.pos, iblockdata != this.state ? 1 : 0);
++                this.level.levelEvent(1500, this.pos, this.composterState != this.state ? 1 : 0);
++                // Paper end
```
Comment (Machine-Maker) [inline patches/server/0739-Add-PlayerCompostItemEvent.patch]: Hmm yeah I forgot about this situation when I was making my example patch...

I don't really like this whole moving the add item to the canPlace... method and would like to leave hopper behavior alone. Taking this into account, having two separate events, EntityCompostItemEvent (is cancellable) and CompostItemEvent (is not cancellable), might actually be better. Can have a note that if they want the hopper input stopped altogether, to use one of the hopper events, not this.

Some more input from others would be good here too.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
A few more things that need to be addressed...

**Comment by NoahvdAa** [comment]
I've now added a parent event, CompostItemEvent. Right now it *only* fires when an item is composted by a hopper, and the EntityCompostItemEvent, as the name implies, only fires when an  entity composts an item. Should the parent event be fired aswell whenever the EntityCompostItemEvent is fired, or is the current implementation fine?

File: patches/api/0327-Add-CompostItemEvent.patch [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
++import org.bukkit.event.HandlerList;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when an item is about to be composted by an entity.
++ */
++public class EntityCompostItemEvent extends CompostItemEvent implements Cancellable {
++
++    private static final HandlerList HANDLER_LIST = new HandlerList();
```
Comment (Machine-Maker) [inline patches/api/0327-Add-CompostItemEvent.patch]: I wouldn't have this event have its own Handler list. I think CompostItemEvent should fire for *all* composts, not just ones from hoppers.

File: patches/server/0739-Add-CompostItemEvent.patch [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
++    static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand, @Nullable Entity entity) {
++        // Paper end
+         // CraftBukkit end
+         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
+         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
+ 
+-        if ((i != 0 || f <= 0.0F) && rand >= (double) f) {
++        // Paper start
++        com.destroystokyo.paper.event.block.CompostItemEvent event = null;
++        com.destroystokyo.paper.event.entity.EntityCompostItemEvent entityEvent = null;
```
Comment (Machine-Maker) [inline patches/server/0739-Add-CompostItemEvent.patch]: Doesn't look like you are using this line

File: patches/api/0327-Add-CompostItemEvent.patch [hunk 12]
```diff
@@ -0,0 +1,139 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Noah van der Aa <ndvdaa@gmail.com>
+Date: Sat, 7 Aug 2021 15:11:27 +0200
+Subject: [PATCH] Add CompostItemEvent.
```
Comment (Machine-Maker) [inline patches/api/0327-Add-CompostItemEvent.patch]:
```suggestion
Subject: [PATCH] Add CompostItemEvent and EntityCompostItemEvent
```

File: patches/server/0739-Add-CompostItemEvent.patch [hunk 13]
```diff
@@ -0,0 +1,103 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Noah van der Aa <ndvdaa@gmail.com>
+Date: Sun, 8 Aug 2021 19:56:02 +0200
+Subject: [PATCH] Add CompostItemEvent.
```
Comment (Machine-Maker) [inline patches/server/0739-Add-CompostItemEvent.patch]:
```suggestion
Subject: [PATCH] Add CompostItemEvent and EntityCompostItemEvent
```

File: patches/server/0739-Add-CompostItemEvent.patch [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+ 
+     static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item) {
+-        // CraftBukkit start
+-        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble());
++        // Paper start
++        return ComposterBlock.addItem(state, world, pos, item, null);
++    }
++    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, @Nullable Entity entity) {
++        // Paper end
++        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
```
Comment (Machine-Maker) [inline patches/server/0739-Add-CompostItemEvent.patch]:
```suggestion
         // CraftBukkit start
         return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble());
     }
+    // Paper start
+    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, @Nullable Entity entity) {
+        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
+    }
+    // Paper end
```

This block of diff can be simplified. (Reminder not to use github's commit from suggestion feature, this is just to show what it should sorta look like)

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

File: patches/api/0327-Add-PlayerCompostItemEvent.patch (hunk 4 above)
Comment (Machine-Maker) [reply patches/api/0327-Add-PlayerCompostItemEvent.patch]: You know, perhaps a getNewLevel()/setNewLevel() pair of methods would be useful. I know right now its only going to increase by one, but make a note in the javadoc for such methods that changing the level on the BlockData will be overriden unless the event is cancelled.

**Comment by Machine-Maker** [comment]
(see https://github.com/PaperMC/Paper/pull/6356#discussion_r685663260)

File: patches/api/0327-Add-CompostItemEvent-and-EntityCompostItemEvent.patch [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when an item is about to be composted by an entity.
++ */
++public class EntityCompostItemEvent extends CompostItemEvent implements Cancellable {
++
++    private Entity who;
++
```
Comment (Machine-Maker) [inline patches/api/0327-Add-CompostItemEvent-and-EntityCompostItemEvent.patch]:
```suggestion
```

Remove this newline between fields

File: patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+     static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item) {
+-        // CraftBukkit start
+-        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble());
++        return ComposterBlock.addItem(state, world, pos, item, null);
++    }
++    // Paper start
++    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, @Nullable Entity entity) {
++        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
+     }
++    // Paper end
```
Comment (Machine-Maker) [inline patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch]: This is still wrong, there is no need to remove the craftbukkit comment or change the line below it

File: patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
++        // Paper end
+         // CraftBukkit end
+         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
+         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
+ 
+-        if ((i != 0 || f <= 0.0F) && rand >= (double) f) {
++        // Paper start
++        com.destroystokyo.paper.event.block.CompostItemEvent event = null;
++        if (generatoraccess != org.bukkit.craftbukkit.util.DummyGeneratorAccess.INSTANCE) {
++            boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f); // taken from if conditional below
```
Comment (Machine-Maker) [inline patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch]: This comment is wrong now since its not really "below" anymore.

```suggestion
+            boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f);
```

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

**Comment by Machine-Maker** [comment]
(These things can be handled by the merger so not really "required changes now")

**Review by Machine-Maker — APPROVED** [review]

**Comment by NoahvdAa** [comment]
Now that I think of it, these events should probably be in the `io.papermc.paper.event` package instead of `com.destroystokyo.paper.event`, right?

**Comment by JRoy** [comment]
> Now that I think of it, these events should probably be in the `io.papermc.paper.event` package instead of `com.destroystokyo.paper.event`, right?

Yes

File: patches/api/0327-Add-CompostItemEvent-and-EntityCompostItemEvent.patch [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
++/**
++ * Called when an item is about to be composted by a hopper.
++ * To prevent hoppers from moving items into composters, cancel the InventoryMoveItemEvent.
++ */
++public class CompostItemEvent extends BlockEvent {
++
++    private static final HandlerList HANDLER_LIST = new HandlerList();
++
++    private ItemStack item;
```
Comment (Machine-Maker) [inline patches/api/0327-Add-CompostItemEvent-and-EntityCompostItemEvent.patch]: final

File: patches/api/0327-Add-CompostItemEvent-and-EntityCompostItemEvent.patch [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
++import org.bukkit.event.Cancellable;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when an item is about to be composted by an entity.
++ */
++public class EntityCompostItemEvent extends CompostItemEvent implements Cancellable {
++
++    private Entity who;
```
Comment (Machine-Maker) [inline patches/api/0327-Add-CompostItemEvent-and-EntityCompostItemEvent.patch]: final

File: patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
++        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
++    }
++    // Paper end
+ 
+     static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand) {
++        // Paper start
++        return ComposterBlock.addItem(iblockdata, generatoraccess, blockposition, itemstack, rand, null);
++    }
++
++    static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand, @Nullable Entity entity) {
```
Comment (Machine-Maker) [inline patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch]: Annotate as Nullable return value

File: patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
++            if (entity == null) {
++                event = new io.papermc.paper.event.block.CompostItemEvent(org.bukkit.craftbukkit.block.CraftBlock.at(generatoraccess, blockposition), itemstack.getBukkitStack(), willRaiseLevel);
++            } else {
++                event = new io.papermc.paper.event.entity.EntityCompostItemEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(generatoraccess, blockposition), itemstack.getBukkitStack(), willRaiseLevel);
++            }
++        }
++
++        if (event != null && !event.callEvent()) {
++            return null;
++        }
```
Comment (Machine-Maker) [inline patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch]: Can be moved inside the DummyGeneratorAccess.INSTANCE conditional check so you don't have to check if the event is null

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

**Comment by NoahvdAa** [comment]
Rebased, comments have also been addressed.

**Review by Machine-Maker — APPROVED** [review]

File: patches/server/0814-Add-CompostItemEvent-and-EntityCompostItemEvent.patch [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
++        // Paper end
+         // CraftBukkit end
+         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
+         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
+ 
+-        if ((i != 0 || f <= 0.0F) && rand >= (double) f) {
++        // Paper start
++        io.papermc.paper.event.block.CompostItemEvent event = null;
++        if (generatoraccess != org.bukkit.craftbukkit.util.DummyGeneratorAccess.INSTANCE) {
++            boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f);
```
Comment (kennytv) [inline patches/server/0814-Add-CompostItemEvent-and-EntityCompostItemEvent.patch]: This changes the outcome of the method for the dummy instance, changing the block used in the entitychangeblockevent

**Review by kennytv — DISMISSED** [review]

**Comment by NoahvdAa** [comment]
Rebased for 1.18.2

File: build-data/mappings-patch.tiny [hunk 23]
```diff
@@ -3,3 +3,9 @@ tiny	2	0	spigot	mojang+yarn
 # Originally DistanceManager, which also implements DistanceManager, so clashes since the implemented class
 # is imported and not fully qualified. Easiest fix is to just change the name
 c	net/minecraft/server/level/PlayerChunkMap$a	net/minecraft/server/level/ChunkMap$ChunkDistanceManager
+
+# Craftbukkit changes the signature of methods in ComposterBlock
+c	net/minecraft/world/level/block/BlockComposter	net/minecraft/world/level/block/ComposterBlock
+	m	(Lnet/minecraft/world/level/block/state/IBlockData;Lnet/minecraft/world/level/World;Lnet/minecraft/core/BlockPosition;Lnet/minecraft/world/entity/Entity;)Lnet/minecraft/world/level/block/state/IBlockData;	d	extractProduce
+	m	(Lnet/minecraft/world/level/block/state/IBlockData;Lnet/minecraft/server/level/WorldServer;Lnet/minecraft/world/item/ItemStack;Lnet/minecraft/core/BlockPosition;Lnet/minecraft/world/entity/Entity;)Lnet/minecraft/world/level/block/state/IBlockData;	a	insertItem
+	m	(Lnet/minecraft/world/level/block/state/IBlockData;Lnet/minecraft/world/level/GeneratorAccess;Lnet/minecraft/core/BlockPosition;Lnet/minecraft/world/item/ItemStack;D)Lnet/minecraft/world/level/block/state/IBlockData;	a	addItem
```
Comment (Machine-Maker) [inline build-data/mappings-patch.tiny]: are these still needed? upstream mapped all the methods and changed the ones they changed.

**Comment by NoahvdAa** [comment]
Rebased and re-tested, on this PRs birthday 🥳 
<br /><br /><br /><br /><br />
~~let's hope it doesn't have a second~~

**Review by Owen1212055 — APPROVED** [review]
